File: backend/services/jwt_service.py
"""JWT access, refresh, and OTP verification tokens."""

from __future__ import annotations

from datetime import datetime, timedelta, timezone
from typing import Any

import jwt

from config.settings import Settings, get_settings
from middleware.errors import ApiError

TOKEN_TYPE_ACCESS = "access"
TOKEN_TYPE_REFRESH = "refresh"
TOKEN_TYPE_OTP = "otp_verification"


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


def _encode(payload: dict[str, Any], settings: Settings | None = None) -> str:
    cfg = settings or get_settings()
    return jwt.encode(payload, cfg.secret_key, algorithm="HS256")


def _decode(token: str, settings: Settings | None = None) -> dict[str, Any]:
    cfg = settings or get_settings()
    try:
        return jwt.decode(token, cfg.secret_key, algorithms=["HS256"])
    except jwt.ExpiredSignatureError as exc:
        raise ApiError("Token has expired", status_code=401, code="TOKEN_EXPIRED") from exc
    except jwt.InvalidTokenError as exc:
        raise ApiError("Invalid token", status_code=401, code="INVALID_TOKEN") from exc


def create_access_token(
    user_id: int,
    role: str,
    *,
    hospital_id: int | None = None,
    district_id: int | None = None,
    settings: Settings | None = None,
) -> str:
    cfg = settings or get_settings()
    expires = _utcnow() + timedelta(minutes=cfg.jwt_access_expires_minutes)
    payload = {
        "sub": str(user_id),
        "role": role,
        "hospital_id": hospital_id,
        "district_id": district_id,
        "type": TOKEN_TYPE_ACCESS,
        "exp": expires,
        "iat": _utcnow(),
    }
    return _encode(payload, cfg)


def create_refresh_token(user_id: int, settings: Settings | None = None) -> tuple[str, datetime]:
    cfg = settings or get_settings()
    expires = _utcnow() + timedelta(days=cfg.jwt_refresh_expires_days)
    payload = {
        "sub": str(user_id),
        "type": TOKEN_TYPE_REFRESH,
        "exp": expires,
        "iat": _utcnow(),
    }
    return _encode(payload, cfg), expires


def create_otp_verification_token(
    phone: str,
    purpose: str,
    settings: Settings | None = None,
) -> str:
    cfg = settings or get_settings()
    expires = _utcnow() + timedelta(minutes=cfg.otp_expires_minutes)
    payload = {
        "phone": phone,
        "purpose": purpose,
        "type": TOKEN_TYPE_OTP,
        "exp": expires,
        "iat": _utcnow(),
    }
    return _encode(payload, cfg)


def decode_token(token: str, *, expected_type: str | None = None) -> dict[str, Any]:
    payload = _decode(token)
    if expected_type and payload.get("type") != expected_type:
        raise ApiError("Invalid token type", status_code=401, code="INVALID_TOKEN_TYPE")
    return payload


def verify_otp_verification_token(token: str, phone: str, purpose: str) -> None:
    payload = decode_token(token, expected_type=TOKEN_TYPE_OTP)
    if payload.get("phone") != phone:
        raise ApiError("OTP verification token does not match phone", status_code=401, code="INVALID_OTP_TOKEN")
    if payload.get("purpose") != purpose:
        raise ApiError("OTP verification token purpose mismatch", status_code=401, code="INVALID_OTP_TOKEN")

File: backend/services/jwtService.js
const jwt = require('jsonwebtoken');
const { ApiError } = require('../middleware/errorHandler');
const crypto = require('crypto');

const JWT_SECRET = process.env.JWT_SECRET;
const JWT_REFRESH_SECRET = process.env.JWT_REFRESH_SECRET;
const JWT_OTP_SECRET = process.env.JWT_OTP_SECRET;

if (!JWT_SECRET || JWT_SECRET.length < 32) {
  console.error('CRITICAL ERROR: JWT_SECRET environment variable is missing or shorter than 32 characters.');
  process.exit(1);
}

if (!JWT_REFRESH_SECRET || JWT_REFRESH_SECRET.length < 32) {
  console.error('CRITICAL ERROR: JWT_REFRESH_SECRET environment variable is missing or shorter than 32 characters.');
  process.exit(1);
}

if (!JWT_OTP_SECRET || JWT_OTP_SECRET.length < 32) {
  console.error('CRITICAL ERROR: JWT_OTP_SECRET environment variable is missing or shorter than 32 characters.');
  process.exit(1);
}

const TOKEN_TYPE_ACCESS = 'access';
const TOKEN_TYPE_REFRESH = 'refresh';
const TOKEN_TYPE_OTP = 'otp_verification';

/**
 * Encodes a payload into a JWT.
 */
function encode(payload, secret) {
  return jwt.sign(payload, secret);
}

/**
 * Decodes and verifies a JWT.
 */
function decodeToken(token, expectedType = null) {
  try {
    const unverified = jwt.decode(token);
    if (!unverified || !unverified.type) {
      throw new ApiError('Invalid token structure', 401, 'INVALID_TOKEN');
    }

    let secret;
    if (unverified.type === TOKEN_TYPE_REFRESH) {
      secret = JWT_REFRESH_SECRET;
    } else if (unverified.type === TOKEN_TYPE_OTP) {
      secret = JWT_OTP_SECRET;
    } else {
      secret = JWT_SECRET;
    }

    const payload = jwt.verify(token, secret);
    if (expectedType && payload.type !== expectedType) {
      throw new ApiError('Invalid token type', 401, 'INVALID_TOKEN_TYPE');
    }
    return payload;
  } catch (error) {
    if (error instanceof ApiError) {
      throw error;
    }
    if (error.name === 'TokenExpiredError') {
      throw new ApiError('Token has expired', 401, 'TOKEN_EXPIRED');
    }
    throw new ApiError('Invalid token', 401, 'INVALID_TOKEN');
  }
}

/**
 * Creates an access token.
 */
function createAccessToken(userId, role, hospitalId = null, districtId = null, tokenVersion = 0) {
  const expiresMinutes = parseInt(process.env.JWT_ACCESS_EXPIRES_MINUTES || '60', 10);
  const payload = {
    sub: String(userId),
    role,
    hospital_id: hospitalId,
    district_id: districtId,
    type: TOKEN_TYPE_ACCESS,
    token_version: tokenVersion,
    exp: Math.floor(Date.now() / 1000) + (expiresMinutes * 60),
    iat: Math.floor(Date.now() / 1000)
  };
  return encode(payload, JWT_SECRET);
}

/**
 * Creates a refresh token and returns the token string and expiry timestamp.
 */
function createRefreshToken(userId) {
  const expiresDays = parseInt(process.env.JWT_REFRESH_EXPIRES_DAYS || '7', 10);
  const durationSeconds = expiresDays * 24 * 60 * 60;
  const expiresAt = new Date(Date.now() + (durationSeconds * 1000));

  const payload = {
    sub: String(userId),
    type: TOKEN_TYPE_REFRESH,
    exp: Math.floor(expiresAt.getTime() / 1000),
    iat: Math.floor(Date.now() / 1000),
    jti: crypto.randomBytes(8).toString('hex')
  };

  return {
    token: encode(payload, JWT_REFRESH_SECRET),
    expiresAt
  };
}

/**
 * Creates an OTP verification token.
 * The `target` can be either a phone number or an email address depending
 * on which channel the OTP was sent through.
 */
function createOtpVerificationToken(target, purpose) {
  const expiresMinutes = parseInt(process.env.OTP_EXPIRES_MINUTES || '5', 10);
  const payload = {
    target,
    purpose,
    type: TOKEN_TYPE_OTP,
    exp: Math.floor(Date.now() / 1000) + (expiresMinutes * 60),
    iat: Math.floor(Date.now() / 1000)
  };
  return encode(payload, JWT_OTP_SECRET);
}

/**
 * Verifies that the OTP verification token matches the target (email or phone) and purpose.
 * Backward-compatible: accepts both `target` (new) and `phone` (legacy) payload fields.
 */
function verifyOtpVerificationToken(token, target, purpose) {
  const payload = decodeToken(token, TOKEN_TYPE_OTP);
  const storedTarget = payload.target || payload.phone; // backward-compat
  if (storedTarget !== target) {
    throw new ApiError('OTP verification token does not match target', 401, 'INVALID_OTP_TOKEN');
  }
  if (payload.purpose !== purpose) {
    throw new ApiError('OTP verification token purpose mismatch', 401, 'INVALID_OTP_TOKEN');
  }
  return true;
}

module.exports = {
  createAccessToken,
  createRefreshToken,
  createOtpVerificationToken,
  decodeToken,
  verifyOtpVerificationToken,
  TOKEN_TYPE_ACCESS,
  TOKEN_TYPE_REFRESH
};
